internals/attachments.py

```diff
@@ -0,0 +1,131 @@
+# Copyright 2024 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License")
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from google.appengine.api import images
+from google.cloud import ndb  # type: ignore
+from typing import Tuple
+
+
+RESIZABLE_MIME_TYPES = [
+    'image/png', 'image/jpg', 'image/jpeg', 'image/gif', 'image/webp',
+    ]
+THUMB_WIDTH = 250
+THUMB_HEIGHT = 200
+MAX_ATTACHMENT_SIZE = 10 * 1024 * 1024  # 10 MB
+
+
+class Attachment(ndb.Model):
+  """Attaches files, such as screenshots, to a feature entry."""
+  feature_id = ndb.IntegerProperty(required=True)
+  created_on = ndb.DateTimeProperty(auto_now_add=True)
+  content = ndb.BlobProperty(required=True)
+  mime_type = ndb.StringProperty(required=True)
+  thumbnail = ndb.BlobProperty()
+  is_deleted = ndb.BooleanProperty(default=False)
+
+
+class UnsupportedMimeType(Exception):
+  pass
+
+class AttachmentTooLarge(Exception):
+  pass
+
+
+_EXTENSION_TO_CTYPE_TABLE = {
+    # These are image types that we trust the browser to display.
+    'gif': 'image/gif',
+    'jpg': 'image/jpeg',
+    'jpeg': 'image/jpeg',
+    'png': 'image/png',
+    'webp': 'image/webp',
+    'txt': 'text/plain',
+}
+
+
+def guess_content_type_from_filename(filename: str) -> str:
+  """Guess a file's content type based on the filename extension."""
+  ext = filename.split('.')[-1] if ('.' in filename) else ''
+  ext = ext.lower()
+  return _EXTENSION_TO_CTYPE_TABLE.get(ext, 'application/octet-stream')
+
+
+def store_attachment(feature_id: int, content: bytes, mime_type: str) -> str:
+  """"Store some data for an attachment.  Return its URI."""
+  check_attachment_size(content)
+  check_attachment_type(mime_type)
+  logging.info('Storing attachment with %r bytes', len(content))
+
+  attachment = Attachment(
+      feature_id=feature_id,
+      content=content,
+      mime_type=mime_type)
+
+  if mime_type in RESIZABLE_MIME_TYPES:
+    # Create and save a thumbnail too.
+    thumb_content = None
+    try:
+      thumb_content = images.resize(content, THUMB_WIDTH, THUMB_HEIGHT)
```

Comment:
Question: Will this code maintain the aspect ratio of the original image if the user uploads an image with different proportions than 250x200? For example, if someone uploads a square image, will it be distorted when resized to 250x200?

Reply:
Nope.  It will be somewhat distorted.  After I get the backend PRs merged, I'll start experimenting with the frontend and coordination with Yann to see how some sample images will look.   My thinking at this point is that all the thumbnails should be the same size so that they look nice when there are multiple, so we might end up with some combination of scaling and cropping.

Reply:
A few quick alternatives to consider in case image dimensions change in the future.

1. Resize with Padding: This will add a white/transparent background around the picture but will keep the original content without distortion
2.  Resize to Fit: Pick the largest dimension and resize it down to its limit (Width=250 or Height=200)

If the images already have the same aspect ratio as 250x200, option 2 is the simplest. These approaches will make our code more robust to future changes.